Added new resolvers related to projectCollaborators

jupiter007 opened the pull request:
## Description

- Added `projectCollaborators` query and `addProjectCollaborators`, `updateProjectCollaborators`, and `removeProjectCollaborators` resolvers to `collaborator.ts` 
- Unfortunately I needed to make direct updates to the data-migrations scripts to remove the creation of `planCollaborators` and just replace with `projectCollaborators` because doing that in subsequent migration scripts was continuously giving me errors.
- I also added `userId` to `projectContributors` in the seeding of the table
- I added a `search` in `Contributor` model to search for users to possibly add as project contributors
- I added a small change to `addProject` to make sure that the user creating that project is added to `projectCollaborators` as an Owner.

Fixes # ([221](https://github.com/CDLUC3/dmsp_backend_prototype/issues/221))

## Type of change
- [X] New feature (non-breaking change which adds functionality)


## How Has This Been Tested?
Manually tested and added a unit test


## Checklist:
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I updated the CHANGELOG.md and added documentation if necessary
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

briri commented on data-migrations/2024-12-17-1133-create-project-tables.sql:
There's no need to make these changes. 
There is a migration that renames `planCollaborators` to `projectCollaborators` [here](https://github.com/CDLUC3/dmsp_backend_prototype/blob/main/data-migrations/2025-02-27-0753-rename-plan-collaborators.sql).

I know we've updated old migration files in the past but we will need to move away from that practice as we start building out the new environments. These types of changes require us to drop the DB and recreate from scratch.

briri reviewed (changes requested):
Lets talk about this one at our stand up. I have reservations about searching the user and contributor/collaborator tables here.

jupiter007 commented:
> Lets talk about this one at our stand up. I have reservations about searching the user and contributor/collaborator tables here.

Yea, that's why I was asking you about it last week. The "Who are the members" page enables users to search for a user based on their name ORCiD or organization and then add them as a member/contributor here: https://www.figma.com/proto/RRtoRLyr5Spf0KiyRfEclt/DMP-Tool-Wireframes?node-id=183-522&t=M9JzcqhvxBJdTW3Q-0&scaling=min-zoom&content-scaling=fixed&page-id=177%3A504&starting-point-node-id=177%3A642

jupiter007 commented on data-migrations/2025-03-31-1200-update-userId-values-in-projectContributors.sql:
```diff
@@ -0,0 +1,15 @@
+UPDATE `projectContributors`
```
@briri I added new data migration scripts instead of updating existing ones, but I was seeing weird behavior.

Sometimes this `UPDATE projectContributors` query would work, and sometimes I'd just get an error saying `Incorrect information in file: './dmsp/projectcontributors.frm'.  

Could you do me a favor and try running these data migrations to see if you get any errors. I'm trying to figure out if its a memory thing since sometimes it works and sometimes it doesn't. Thanks.

briri replied:
Will do. Those `Incorrect information in file: '*.frm` errors are I think due to the info stored on disk becoming out of sync with the MariaDB instance that is running. It happens sometimes when you run migrations and don't restart the container. Usually doing a `docker-compose down` followed by a `docker-compose up` fixes the issue for me

briri commented on data-migrations/2025-03-31-1025-add-userId-to-projectContributors-table.sql:
we don't need the `userId` on this table. Contributors will most likely not have an account, so no user record

jupiter007 replied:
@briri the reason I added userId to the projectContributors table was that I needed a way to tie in projectContributors with the access levels you added to the projectCollaborators table for the Project Members page here: https://www.figma.com/proto/RRtoRLyr5Spf0KiyRfEclt/DMP-Tool-Wireframes?node-id=183-612&t=M9JzcqhvxBJdTW3Q-0&scaling=min-zoom&content-scaling=fixed&page-id=177%3A504&starting-point-node-id=177%3A642. 

Any suggestions on how you would tie them together?

briri commented on src/schemas/contributor.ts:
```diff
@@ -47,6 +47,8 @@ export const typeDefs = gql`
     project: Project
     "The contributor's affiliation"
     affiliation: Affiliation
+    "The user id of the contributor"
```
don't need this on the contributor record

jupiter007 commented on src/models/Contributor.ts:
```diff
@@ -7,6 +7,7 @@ import { MySqlModel } from "./MySqlModel";
 export class ProjectContributor extends MySqlModel {
   public projectId: number;
   public affiliationId?: string;
+  public userId: number;
```
@briri I added the `userId` field because I needed a way to tie projectContributors to the `access levels` you added in `projectCollaborators`.

jupiter007 commented:
Converting this to `Draft` mode until we figure out a good way to tie `projectContributors` to the `access levels` in the `projectCollaborators` table.

jupiter007 commented:
Hi @briri I updated this code to just focus on the `projectCollaborators` resolvers, and removed any updates I made related to `search`. There is a separate ticket created for that: #251

briri commented on src/resolvers/collaborator.ts:
```diff
@@ -101,12 +123,136 @@ export const resolvers: Resolvers = {
+      const reference = 'addProjectCollaborator resolver';
+      try {
+        // if the user is an admin
+        if (isAdmin(context.token)) {
```
Same as above. Collaborators with 'EDIT' or 'OWN' access on the project should be able to do this, so the `hasPermissionOnProject` check will be enough. No need to check if they are an admin

briri commented on src/resolvers/collaborator.ts:
```diff
@@ -101,12 +123,136 @@ export const resolvers: Resolvers = {
+      const reference = 'updateProjectCollaborator resolver';
+      try {
+        // if the user is an admin
+        if (isAdmin(context.token)) {
```
remove this check

briri commented on src/resolvers/collaborator.ts:
```diff
@@ -101,12 +123,136 @@ export const resolvers: Resolvers = {
+      const reference = 'removeProjectCollaborator resolver';
+      try {
+        // if the user is an admin
+        if (isAdmin(context.token)) {
```
and this one

briri reviewed (changes requested):
Looks good @jupiter007. The only change is to remove the `isAdmin` checks, a researcher will need to be able to modify projectCollaborators.

briri reviewed (approved):
thanks @jupiter007 LGTM
